### PR TITLE
Youtube fixes

### LINKED ISSUE #711: Youtube: Cannot download video
For two days now, more than half of the youtube channels I subscribe to are not downloadable. The problem is that their feeds don't contain 'url_encoded_fmt_stream_map'. The videos can be downloaded with youtube-dl and enabling the youtube-dl extension allows them to be downloaded in gpodder. I don't like using the extension because it takes forever to update, and will be updating with gpodder and then temporarily enabling the extension to download. I am looking for a fix...

### LINKED ISSUE #714: [Youtube] should gpodder fail when preferred formats aren't available?
In youtube.py:
```
# Default to the highest fmt_id if we don't find a match below
_, url = fmt_id_url_map[0]
```

That selects the format with the highest value when a preferred format is not available. It basically grabs a random format since they aren't ordered based on video codec, resolution or quality. Gpodder has a list of fallback formats to step down through video sizes for each codec.

If you only want MP4 content, the random chooser would probably grab format 43 (WebM) if none of the MP4 formats were available.

I set youtube.preferred_fmt_ids thinking it would keep gpodder from downloading tiny videos when larger ones weren't available. But that random chooser kicks in grabs something.

```
else:    
    raise YouTubeError('No preferred formats found for video ID "%s"' % vid)    
```

Adding that below the for loop causes gpodder to fail when a preferred a format isn't available, instead of choosing a random format. So, should gpodder randomly choose a format you probably don't want, or should it fail?

### LINKED ISSUE #719: Support adaptive formats for youtube-dl extension
http://borg.qzx.com/files/gpodder-719-support-adaptive-formats-for-youtube-dl-extension.patch

This patch adds adaptive formats to the default format lists for use by the youtube-dl extension. It skips over them when downloading with gpodder's code.

It only adds 22 -> 136+140 and 18 -> 134+140, but more can be added.

### PR DESCRIPTION
**Pull request by elelay**
multiple fixes for Youtube.
Fixes #711 
Fixes #714 
Fixes #719

### CODE REVIEW

File: src/gpodder/youtube.py [hunk 1]
```diff
@@ -44,33 +44,33 @@
     # Fallback to an MP4 version of same quality.
     # Try 34 (FLV 360p H.264 AAC) if 18 (MP4 360p) fails.
     # Fallback to 6 or 5 (FLV Sorenson H.263 MP3) if all fails.
-    (46, ([46, 37, 45, 22, 44, 35, 43, 18, 6, 34, 5],
+    (46, ([46, 37, 45, 22, '136+140', 44, 35, 43, 18, '134+140', 6, 34, 5],
```
Comment (elelay) [inline src/gpodder/youtube.py]: The same format shouldn't appear multiple time, should it?

Comment (auouymous) [reply src/gpodder/youtube.py]: They aren't the same formats, one is 13-SIX and the other is 13-FOUR. 136 is the video for 22 and 134 is video for 18, both use the same audio.

**Comment by auouymous** [comment]
Looks good to me. Are the two build failures a problem with Travis CI or something elsee we need to fix first?

**Comment by elelay** [comment]
> Looks good to me. Are the two build failures a problem with Travis CI or something else we need to fix first?
you're right: my mistake

File: src/gpodder/youtube.py (hunk 1 above)
Comment (elelay) [reply src/gpodder/youtube.py]: you're right: my mistake

**Comment by elelay** [comment]
> Looks good to me. Are the two build failures a problem with Travis CI or something else we need to fix first?

Doesn't seem related. `make releasetest` is what is run.

**Comment by elelay** [comment]
Indeed: it was an update of the coverage package that broke the ci, also on master.

**Comment by elelay** [comment]
@auouymous I've adapted your code from #718 with a new downloader option for the DownloadTask Can you please test and review it?
Thanks,

**Comment by auouymous** [comment]
Works great, thanks!

Why are the other two function calls needed in gtkui/main.py on_download_selected_episodes(), but download-with-ytdl seems to work fine without them?

```
self.gpodder.update_episode_list_icons([episode.url for episode in episodes])
self.gpodder.play_or_download()
```

**Comment by elelay** [comment]
They must be leftover from when the model wasn't updated automatically?? When clicking "Download selected episodes" from the new episodes dialog it's not called. Neither from the download episode function given to extensions. I've removed these 2 lines in bc1e735.

**Comment by elelay** [comment]
Merged, thanks!

**Comment by luckybuzz** [comment]
@elelay @auouymous 

very sorry to jump in with such a basic question, but after spending 1.5 hours breaking things on my own, i thought it can't hurt to ask for help.

i'm running a windows installation of gpodder (used the 3.10.11 installer exe).  in hopes of stopping the youtube error messages, i downloaded each of the 4 changed files here and replaced the matching files on my computer with them.  

clearly that wasn't the right thing to do as now nothing downloads at all and i'm getting different python error messages (missing attributes and local variables referenced before assignment, etc.)

is there a way i can apply these fixes to my Windows installation of gpodder?  or do i just need to sit tight and wait for the next release?

**Comment by elelay** [comment]
@luckybuzz you can reinstall using the installer from the continuous build: https://ci.appveyor.com/project/elelay/gpodder/builds/29953936/artifacts
Hope this helps...

**Comment by luckybuzz** [comment]
it really does @elelay thank you!  i didn't realize a continuous build was available.  cheers  :)

**Comment by luckybuzz** [comment]
@elelay regrettably, i'm still getting the error messages after installing from the continuous build exe you linked above.  not sure if the problem is on my computer or in the code, but in case it helps you in any way, here is the error traceback from my gpodder logs:

2020-01-10 16:09:30,284 [gpodder.download] ERROR: Download failed: local variable 'error_message' referenced before assignment
Traceback (most recent call last):
  File "C:\Program Files (x86)\gPodder\lib\python3.8\site-packages\gpodder\download.py", line 861, in run
    downloader = DefaultDownloader.custom_downloader(self._config, self.episode)
  File "C:\Program Files (x86)\gPodder\lib\python3.8\site-packages\gpodder\download.py", line 431, in custom_downloader
    res = registry.download_url.resolve(config, None, episode)
  File "C:\Program Files (x86)\gPodder\lib\python3.8\site-packages\gpodder\registry.py", line 31, in resolve
    result = resolver(item, *args)
  File "C:\Program Files (x86)\gPodder\lib\python3.8\site-packages\gpodder\youtube.py", line 120, in youtube_real_download_url
    res, duration = get_real_download_url(episode.url, fmt_ids)
  File "C:\Program Files (x86)\gPodder\lib\python3.8\site-packages\gpodder\youtube.py", line 187, in get_real_download_url
    fmt_id_url_map = sorted(find_urls(page), reverse=True)
  File "C:\Program Files (x86)\gPodder\lib\python3.8\site-packages\gpodder\youtube.py", line 177, in find_urls
    if error_message:
UnboundLocalError: local variable 'error_message' referenced before assignment

**Comment by auouymous** [comment]
Does the problem go away if you add `error_message = None` after the `x = parse_qs(page)` line in youtube.py?

**Comment by luckybuzz** [comment]
hi @auouymous!  no, not only does the problem not go away... gpodder itself won't start anymore!

i tried adding and taking away that line of code from youtube.py multiple times.  when it's not there, gpodder runs, but when i add it, the executable does nothing.  very strange!

**Comment by auouymous** [comment]
Did you add the same number of spaces (indent) before it as the existing line?

**Comment by luckybuzz** [comment]
@auouymous i did indeed.

after a bit of googling, i tried another solution.  i made error_message the second variable of the find_urls function, and then passed in value 'None' in line 187.

that allowed gpodder to run, but now when i try to download any video it's giving me the error message from line 193 (No formats found for video ID...)

**Comment by auouymous** [comment]
https://github.com/gpodder/gpodder/blob/master/src/gpodder/youtube.py

Your line numbers don't match up with those in the above url. The no-formats message would be line 196 if you added a new line for the error_message variable. I'm not even sure what the line 187 is you changed.

**Comment by luckybuzz** [comment]
@auouymous oh i see what you mean.  the youtube.py that installed from the exe linked above by @elelay is a few lines shorter than the one you just posted.

where yours introduces variable playabilityStatus in line 158 and then uses that variable in line 160. 

the code in youtube.py i'm using has no line playabilityStatus variable and has a few other differences in the next few lines of code.

i tried substituting my youtube.py with yours, and now i'm back to gpodder not loading at all.

very strange stuff.  thank you for your attempts to help  :)

**Comment by auouymous** [comment]
https://ci.appveyor.com/project/elelay/gpodder/builds/30005447/artifacts

That is the most current build and should have the correct line numbers. Try making it look like this.

```
            x = parse_qs(page)
            error_message = None

            ...

            if error_message is not None:
                raise YouTubeError('Cannot download video: %s' % error_message)
```

**Comment by luckybuzz** [comment]
it looks like that problem is now resolved, because the local variable error message is gone.  thank you  :)

unfortunately, now every video i attempt to download gives the "no format found" error.  i will keep experimenting over the weekend, but here is the error traceback if you are curious.  this is just for the most recent video i tried, but it's the same on all of them.

2020-01-10 18:34:25,332 [gpodder.download] ERROR: Download failed: No formats found for video ID "B1tdcS_3qYk"
Traceback (most recent call last):
  File "C:\Program Files (x86)\gPodder\lib\python3.8\site-packages\gpodder\download.py", line 861, in run
    downloader = DefaultDownloader.custom_downloader(self._config, self.episode)
  File "C:\Program Files (x86)\gPodder\lib\python3.8\site-packages\gpodder\download.py", line 431, in custom_downloader
    res = registry.download_url.resolve(config, None, episode)
  File "C:\Program Files (x86)\gPodder\lib\python3.8\site-packages\gpodder\registry.py", line 31, in resolve
    result = resolver(item, *args)
  File "C:\Program Files (x86)\gPodder\lib\python3.8\site-packages\gpodder\youtube.py", line 120, in youtube_real_download_url
    res, duration = get_real_download_url(episode.url, fmt_ids)
  File "C:\Program Files (x86)\gPodder\lib\python3.8\site-packages\gpodder\youtube.py", line 196, in get_real_download_url
    raise YouTubeError('No formats found for video ID "%s"' % vid)
gpodder.youtube.YouTubeError: No formats found for video ID "B1tdcS_3qYk"

**Comment by auouymous** [comment]
Changing those two lines fixed the error_message problem?

Can you insert prints at the following lines and tell me which ones gpodder outputs when downloading a video?

```
            elif 'player_response' in x:
                print('P')

                elif 'streamingData' in player_response:
                    print('S')

                    if 'formats' in player_response['streamingData']:
                        print('F')

            print('Q')
            r4 = re.search('url_encoded_fmt_stream_map=([^&]+)', page)
            if r4 is not None:
                print('R')
```

**Comment by luckybuzz** [comment]
hi @auouymous.  out of the five, only Q is printing on download attempts.

**Comment by auouymous** [comment]
Put this raise line after the error_message declaration. What does it display when you try to download?

```
            error_message = None
            raise YouTubeError(page)
```

**Comment by luckybuzz** [comment]
hmm it gave me the very generic "Failed: Error."

2020-01-10 22:55:47,775 [gpodder.download] ERROR: Download failed: 
Traceback (most recent call last):
  File "C:\Program Files (x86)\gPodder\lib\python3.8\site-packages\gpodder\download.py", line 861, in run
    downloader = DefaultDownloader.custom_downloader(self._config, self.episode)
  File "C:\Program Files (x86)\gPodder\lib\python3.8\site-packages\gpodder\download.py", line 431, in custom_downloader
    res = registry.download_url.resolve(config, None, episode)
  File "C:\Program Files (x86)\gPodder\lib\python3.8\site-packages\gpodder\registry.py", line 31, in resolve
    result = resolver(item, *args)
  File "C:\Program Files (x86)\gPodder\lib\python3.8\site-packages\gpodder\youtube.py", line 120, in youtube_real_download_url
    res, duration = get_real_download_url(episode.url, fmt_ids)
  File "C:\Program Files (x86)\gPodder\lib\python3.8\site-packages\gpodder\youtube.py", line 196, in get_real_download_url
    fmt_id_url_map = sorted(find_urls(page), reverse=True)
  File "C:\Program Files (x86)\gPodder\lib\python3.8\site-packages\gpodder\youtube.py", line 154, in find_urls
    raise YouTubeError(page)
gpodder.youtube.YouTubeError``

**Comment by auouymous** [comment]
What about when moving the raise before `page = page.decode()`?

**Comment by luckybuzz** [comment]
that's different.  it gave me Error: b''

2020-01-10 23:05:16,000 [gpodder.download] ERROR: Download failed: b''
Traceback (most recent call last):
  File "C:\Program Files (x86)\gPodder\lib\python3.8\site-packages\gpodder\download.py", line 861, in run
    downloader = DefaultDownloader.custom_downloader(self._config, self.episode)
  File "C:\Program Files (x86)\gPodder\lib\python3.8\site-packages\gpodder\download.py", line 431, in custom_downloader
    res = registry.download_url.resolve(config, None, episode)
  File "C:\Program Files (x86)\gPodder\lib\python3.8\site-packages\gpodder\registry.py", line 31, in resolve
    result = resolver(item, *args)
  File "C:\Program Files (x86)\gPodder\lib\python3.8\site-packages\gpodder\youtube.py", line 120, in youtube_real_download_url
    res, duration = get_real_download_url(episode.url, fmt_ids)
  File "C:\Program Files (x86)\gPodder\lib\python3.8\site-packages\gpodder\youtube.py", line 144, in get_real_download_url
    raise YouTubeError(page)
gpodder.youtube.YouTubeError: b''

**Comment by auouymous** [comment]
Replace 'page' in the raise line with 'url'.

**Comment by luckybuzz** [comment]
that one just says "Error." in the GUI but log says the following:

2020-01-10 23:18:22,083 [gpodder.download] ERROR: Download failed: https://www.youtube.com/get_video_info?&el=detailpage&video_id=B1tdcS_3qYk
Traceback (most recent call last):
  File "C:\Program Files (x86)\gPodder\lib\python3.8\site-packages\gpodder\download.py", line 861, in run
    downloader = DefaultDownloader.custom_downloader(self._config, self.episode)
  File "C:\Program Files (x86)\gPodder\lib\python3.8\site-packages\gpodder\download.py", line 431, in custom_downloader
    res = registry.download_url.resolve(config, None, episode)
  File "C:\Program Files (x86)\gPodder\lib\python3.8\site-packages\gpodder\registry.py", line 31, in resolve
    result = resolver(item, *args)
  File "C:\Program Files (x86)\gPodder\lib\python3.8\site-packages\gpodder\youtube.py", line 120, in youtube_real_download_url
    res, duration = get_real_download_url(episode.url, fmt_ids)
  File "C:\Program Files (x86)\gPodder\lib\python3.8\site-packages\gpodder\youtube.py", line 144, in get_real_download_url
    raise YouTubeError(url)
gpodder.youtube.YouTubeError: https://www.youtube.com/get_video_info?&el=detailpage&video_id=B1tdcS_3qYk

**Comment by auouymous** [comment]
Okay, the `page = req.read()` line is returning an empty string. And it isn't being redirecting to some bogus url. Can gpodder download from sites other than youtube? Are you in control of the python 3.8 installation or did gpodder install that?

I have no idea what it could be if feeds other than youtube work. If nothing works, it might be an issue in util.py http_request(), and maybe caused by python 3.8.

**Pull request merged**